### ISSUE TITLE
Some points regarding the whatlies paper

### ISSUE THREAD

**Issue by mkaze** [RasaHQ/whatlies#209]
I just finished reading the [whatlies paper](https://arxiv.org/abs/2009.02113). Very well done and written! :+1: 

However, there were some points which I would like to mention here. Specifically, there are some inconsistencies between paper and what has been implemented in whatlies which I think are worth addressing if the authors wanted to update the pre-print on arXiv or submit it to a conference/journal in future, or even change the implementation instead.

---
### Sentence representation
There are some problems in the following paragraph (p.2, right column, at the top):

> In order to retrieve a sentence representation for word-level embeddings such as fastText, whatlies returns the summed representation of the individual word vectors. For pre-trained encoders such as BERT (Devlin et al., 2019) or ConveRT (Henderson et al., 2019), whatlies uses its internal [CLS] token for representing a sentence.

This is actually more or less incorrect, because:
1. The current implementation of `FasttextLanguage` uses `get_word_vector` method of the `fasttext` library: https://github.com/RasaHQ/whatlies/blob/6811b7a24b71f169c5b97a0f8a17572f07f96f94/whatlies/language/_fasttext_lang.py#L105-L109 This method first extracts n-grams from the given string (it doesn't check if it's a single or multiple tokens) and returns the **average** of n-gram embeddings, no matter it's a single token or a sentence ([source code](https://github.com/facebookresearch/fastText/blob/2cc7f54ac034ae320a9af784b8145c50cc68965c/src/fasttext.cc#L111-L120)). Further, I would argue that it would have been better to use `get_sentence_vector` instead which consumes the string token by token, compute the representation of each token (as I just explained), and then returns the **average of l2-normalized** token embeddings ([source code](https://github.com/facebookresearch/fastText/blob/2cc7f54ac034ae320a9af784b8145c50cc68965c/src/fasttext.cc#L502-L518)). If the given query was a single token, then this function would work correctly in that case as well (though, the downside is that it would be l2-normalized which may not be desirable).

2. I don't know about ConveRT, but regarding the huggingface transformers in general: first of all, not all transformers have a `[CLS]` token or its equivanlent (e.g. GPT, GPT2). Secondly, we are not using `[CLS]` token representation in `whatlies`; rather, the **sum** of non-special (wordpiece) tokens are used as the representation of the given query: https://github.com/RasaHQ/whatlies/blob/6811b7a24b71f169c5b97a0f8a17572f07f96f94/whatlies/language/_hftransformers_lang.py#L80-L86 This is actually in consistency with the `spacy-transformers` library where the sum of token representations is used as `doc.vector` (though, there is an alignment process involved in `spacy-transformers` to align wordpieces back to spaCy tokens).

3. spaCy uses the **average** token embeddings as the representation of `doc.vector`.

4. TF Hub text embedding models usually combine token embeddings using `sqrtn` weighting to form the sentence representation ([documentation](https://www.tensorflow.org/api_docs/python/tf/nn/embedding_lookup_sparse#args); see `combiner`).

Further, there is this comment (p.3, right column, first code snippet) which is a bit incorrect according the explanation above:

> \# NB: 'sentences also embed' will be represented as the sum of the 3 individual words.

It's "average", not sum, to be accurate.

---
### Usage of helper vectors

Actually, since #205 has been merged, there is no longer any helper vectors. However, they are used in the paper (p.4, right column, "Transformations" section). Further, in the same section of the paper, the explanation after the code snippet is no longer valid and relevant (and also a bit misleading as I explained [here](https://github.com/RasaHQ/whatlies/issues/207#issuecomment-689553684)).

---
That's it! The paper was very well-written and clear, so I couldn't find any further problem :smile: :smile: I just have a few minor suggestions which I would write later in a separate comment of this issue.

**Comment by koaning** [comment]
The miracle of open source is that things can get dated quite quickly. 

This is grand feedback, thanks @mkaze ! 

We haven't gotten feedback from the review committee yet but once that is in we'll make sure that @tttthomasssss and I adress the issues that you mention here. In particular;

- I'll adress the code snippets that contain the "helper vectors". It should make the code shorter. 
- We'll also change the comment there. We'll need to mention that different backends currently have a different method of pooling all the word vectors in the sentence and that it is currently inconsistent in our library. This also gives us to opportunity to mention this as future work.

**Comment by mkaze** [comment]
You are very welcome! Yeah, these are not major problems, per say; however, I thought it would be good to mention them, so you might be interested to address them in future.

There are also just a few suggestions from my side; though, you may not necessarily find them useful :smile: Here they are:

1. The code snippets and the vector arithmetic operators are very clear, and therefore even someone who has never used the library could easily understand them to a great extent. However, there is one exception: the projection operator, i.e. `|`. Actually, it has been used several times in the code snippets but you have not mentioned or defined it in the paper (or maybe I have missed it?!). I am not sure if this is intentional or not, but by just reading the codes this may not be clear for the reader that what this operator is doing exactly. I think it's worth to mention this operator or define it, the same way you have defined scalar projection in page 4.

2. Isn't it better to annotate the points of each intent class in Figure 8 with a different color so as to show that the clusters has been formed correctly in ConveRT plot according to the classes, or the opposite for the spaCy plot? As you are aware, this could be done by using `add_property` and `plot_interactive`  methods. 

3. In page 4 where the projection for visualizations is explained it might be worth mentioning that the custom axis metrics other than the default scalar projection could be provided as well. Of course, this is a newly-added feature as well, and also the length of the paper is limited!

That's it for now! :smile: Good luck!

**Comment by koaning** [comment]
It was a bit rush-rush at the end but we've adressed some of these notes while making the final draft. Will close for now.
